packages/next-codemod/bin/upgrade.ts

```diff
@@ -134,13 +140,21 @@ export async function runUpgrade(
     // x-ref(PR): https://github.com/vercel/next.js/pull/65058
     // x-ref(release): https://github.com/vercel/next.js/releases/tag/v14.3.0-canary.45
     compareVersions(targetNextVersion, '14.3.0-canary.45') >= 0 &&
-    installedReactVersion.startsWith('18')
+    installedReactVersion.startsWith('18') &&
+    // Pure App Router always uses React 19
+    // The mixed case is tricky to handle from a types perspective.
+    // We'll recommend to upgrade in the prompt but users can decide to try 18.
+    !isPureAppRouter
   ) {
     const shouldStayOnReact18Res = await prompts(
       {
         type: 'confirm',
         name: 'shouldStayOnReact18',
-        message: `Are you using ${pc.underline('only the Pages Router')} (no App Router) and prefer to stay on React 18?`,
+        message:
+          `Do you prefer to stay on React 18?` +
+          (isMixedApp
+            ? " Since you're using both pages/ and app/, we recommend upgrading React to use a consistent version throughout your app."
```

Comment:
Staying on 18 is also consistent version, maybe inform (benefit of | risk of not) upgrading?

Reply:
> Staying on 18 is also consistent version

Not if you're using App Router (always uses React 19) which should the case in this branch.

We should link to the upgrade guide earlier in the CLI regardless. This is not the place to go into details about this. Or do we link to the upgrade guide at the end?

Reply:
I see. No I think is good for how it is.